src/middle/Validation_intf.ml

```diff
@@ -0,0 +1,42 @@
+module type Infix = sig
+  type 'a t
+
+  val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
```

Comment:
There's part of me that really likes abstracting out this sort of validation monad because obviously that is what is going on with this sort of error accumulation. Another part of me wonders whether it's slightly overengineered and would make the semantic check harder to understand for new developers and harder to maintain.
Let me think about it a bit longer. I think I mostly like it.

Reply:
Yeah, I get where you're coming from. 

For type-checking / semantic-checking, what we actually want is a monad transfomer stack with `Reader` for the options, `Validation` to accumulate errors and `State` to carry the environment around.

Clearly this is going to be quite intimidating for most developers and probably impenetrable for someone with out an FP background. 

On the plus side, I this the approach will make something like a language server easier and should make the experience nicer for users i.e. we can report syntactic and semantic errors together rather than just the first.

Reply:
With all that being said, I think we can actually make the API to the `SemanticCheck` monad quite nice so that beginners can almost treat it like a DSL.

Reply:
OK. I've thought about it more and don't see a way around it. I think this is more or less the most slick and least intimidating way of achieving what we want.